### ISSUE TITLE
Slow web page loading

### ISSUE THREAD

**Issue by mihem** [iSEE/iSEE#459]
iSEE is a great interactive visualization tool that allows easily integration with common pipelines such as Seurat, SingleCellExperimens and allows to put these visualization on your own server easily using ShinyServer. I think the main flaw is the time it takes to load the webpage. For an average dataset it took me around 20s, for a large single cell RNA seq data set I took me around 90s to load the page. Also see this nice comparison https://www.biorxiv.org/content/10.1101/2020.01.24.918342v2.full, which identifies loading of the web page as one main flaw of ISEE compared to other tools. After loading the webpage  the app is actually really fast. 

Do you have any current solutions or ideas for the future how to speed up this long preprocessing time?

**Comment by LTLA** [comment]
There's a couple of points here.

The first is how you're setting up your Shiny Server. I'll admit that I actually don't remember the config options because I mostly use Rstudio Connect nowadays, but I'll guess that the default setting probably does not use a minimum number of processes. This means that anyone going to your Shiny server has to wait for the R session to start and load in all of **iSEE**'s dependencies (probably a big part of your 20 seconds) and then load in all of the data (probably the extra 70 seconds for the larger single-cell datasets). 

So one solution is to specify a minimum number of processes. This means that the R package loading and data loading is done once when the process starts, and then users only have to wait a few seconds for the set-up of the Shiny session. As I said, I don't know how to do this for the Shiny Server, but I know that Rstudio Connect has a "minimum number of processes" option, and I usually set this to 1 for lightly used apps and maybe 3-5 for for apps that I might expect to have more traffic. (You don't want to specify too many minimum processes because those will use up your server resources.) Users beyond the minimum number of processes will then experience a delay as the R session boots up.

Another complementary solution is to use a disk-backed format for your large datasets. I would suggest having a look at the **HDF5Array** package where you can store matrix data in a HDF5 file and bits of it are retrieved on demand. In the context of **iSEE**, we generally only need individual rows or columns of the matrix (e.g., to color cells by an expression of the gene) and that can be retrieved from file very quickly. The major advantage is that the data loading time is effectively eliminated because the entire dataset never needs to be loaded into memory when the app starts. 

Beyond that, there are some lesser tricks to cut down the Shiny loading time. This includes downsampling points, see `?iSEEOptions` (though note that these only take effect _before_ construction of the panels); and initializing the app with only the panels that you actually need (no need to spend time rendering plots that you don't care about).

**Comment by mihem** [comment]
Thanks a lot Aaron for the detailed answer!

Unfortunately, I couldn`t find a similar setting as "minimum number of processes" in the free Shiny Server. In the cited publication it says: 

> R processes for faster load times (https://rstudio.com/products/shiny/shiny-server/) and therefore it starts a new R process for each new user, increasing the loading times and affecting the user experience when data sets are large.

So I guess this is just not an option if your institution doesn't have R Studio Connect. Or do you know any workaround?

But when setting `iSEEOptions`  I realized that I still had the old iSEE1.6 version running since Ubuntu LTS default ist still R 3.6 and `iSEEOptions` is not available in that version. Only updating brought a dramatic decrease of loading the webpage (from like 90s to around 30sec for the large scRNA seq dataset with around 60 000 cells). I wonder why ...

Using HDF5Array also reduced loading time quite a bit. Reducing the number of plots to the very essential ones also really helpe. Downsampling was less effective in my case and also reduced biological significance, therefore I did not that in the final version.

To sum up for any other user having the same problems, I would recommend (thanks Aaron!):

1) Update!!
2) Use HDF5Array
3) Reduce the number of plots

@Aaron Maybe it's wortwhile including these tips in your FAQ or some other place of the vignette.

**Comment by kevinrue** [comment]
Hi @mihem 

Thanks for the question.

The documentation for working with large datasets in currently on devel.
https://bioconductor.org/packages/devel/bioc/vignettes/iSEE/inst/doc/bigdata.html

It will be part of the October release of Bioconductor.

Note that you can also also find the 'devel' version of the other vignettes here:
https://bioconductor.org/packages/devel/bioc/html/iSEE.html

Best
Kevin

**Comment by mihem** [comment]
Hey Kevin, thanks, that vignette is really helpful.
Great work!
